nano/lib/signal_manager.cpp

```diff
@@ -0,0 +1,101 @@
+#include <nano/lib/signal_manager.hpp>
+#include <nano/lib/utility.hpp>
+
+#include <boost/asio.hpp>
+#include <boost/asio/signal_set.hpp>
+#include <boost/bind.hpp>
+#include <boost/format.hpp>
+
+#include <iostream>
+
+nano::signal_manager::signal_manager () :
+	work (boost::asio::make_work_guard (ioc))
+{
+	smthread = boost::thread ([&ioc = ioc] () {
+		ioc.run ();
+	});
+}
+
+nano::signal_manager::~signal_manager ()
+{
+	/// Indicate that we have finished with the private io_context. Its
+	/// io_context::run() function will exit once all other work has completed.
+	work.reset ();
+	ioc.stop ();
+	smthread.join ();
+}
+
+nano::signal_manager::signal_descriptor::signal_descriptor (std::shared_ptr<boost::asio::signal_set> sigset_a, signal_manager & sigman_a, std::function<void (int)> handler_func_a, bool repeat_a) :
+	sigset (sigset_a), sigman (sigman_a), handler_func (handler_func_a), repeat (repeat_a)
+{
+}
+
+void nano::signal_manager::register_signal_handler (int signum, std::function<void (int)> handler, bool repeat)
+{
+	// create a signal set to hold the mapping between signals and signal handlers
+	std::shared_ptr<boost::asio::signal_set> sigset (new boost::asio::signal_set (ioc, signum));
+
+	// a signal descriptor holds all the data needed by the base handler including the signal set
+	// working with copies of a descriptor is OK
+	signal_descriptor descriptor (sigset, *this, handler, repeat);
+
+	// ensure the signal set and descriptors live long enough
+	descriptor_list.push_back (descriptor);
+
+	// asynchronously listen for signals from this signal set
+	sigset->async_wait (boost::bind (nano::signal_manager::base_handler, descriptor, _1, _2));
+
+	log (boost::str (boost::format ("Registered signal handler for signal %d") % signum));
+}
+
+void nano::signal_manager::set_logger (nano::logger_mt * logger_a)
+{
+	logger = logger_a;
+}
+
+std::function<void (int)> nano::signal_manager::get_debug_files_handler (void)
+{
+	return [] (int) {
+		nano::dump_crash_stacktrace ();
+		nano::create_load_memory_address_files ();
+	};
+}
+
+void nano::signal_manager::log (const std::string msg)
```

Comment:
The string should be passed by reference otherwise its copied when calling this function.

Since const can be used in multiple points in a type declaration, we stick it to the right of the thing we're consting:

void nano::signal_manager::log (std::string const & msg)